[Do not merge] Ideas how to use apq

elsid opened the pull request.

elsid committed:
First simple ideas how to use

elsid committed:
Add column name for binary protocol

elsid committed:
Ideas how to use config and timeouts

elsid committed:
Use cursor to receive data

elsid committed:
Binary protocol without defined struct

elsid committed:
Named query variables

Commit:
Add some interfaces prototypes

And correct a couple of examples with cursor. The main idea of fetch is to do not
expose any interface of the cursor and allow the same semantic as request for a
whole result.

thed636 commented on ideas/single_connection_usage_with_tuple_cursor.cpp:
```diff
@@ -0,0 +1,12 @@
+    apq::cursor<std::tuple<std::string>> cursor;
+    apq::binary_protocol::request(connection_provider, cursor, "SELECT pg_is_in_recovery()", yield);
+    std::tuple<std::string> data;
+    apq::binary_protocol::request(cursor, data, yield);
```
Here must be ```fetch``` of course

elsid commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+*   std::function<void(PGconn*)> - finalization function, e.g. destructor for single 
+*                                  connection or recycling operation for pool.
+*/
+using pg_connect_handle = std::shared_ptr<PGconn>; //, std::function<void(PGconn*)>>;
```
I think we definitely shouldn't use ```std::function``` for finalization. For both standalone and polled connection we know finalization functions. Let's use some customized type or two different types.

thed636 replied:
There is a trade-off between std::function (or any other abstraction) and several types. It is about connection class. In case of the different types we need to use a template of the connection class. And this leads to templating of the transaction_connection class, commit and rollback functions. IMO this fact makes code slightly more complicated. But I can be wrong.

elsid commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+* is not attractive. We can provide some kind of yield_context support, but with
+* message loss in case of error_code bound to yield (i.e. yield[ec])
+*/
+using error_code = boost::system::error_code; // std:: / custom ?
```
Until we use ```boost::asio``` let's use ```boost::system_error```. For networking ts we'll back to this question.

thed636 replied:
Well, it doesn't matter which standard code we use std or boost, the main problem is when we use standard error_code we loose an error message form PostgreSQL. We can use custom error_code like this prototype in my repo https://github.com/thed636/error_code/blob/master/include/error_code/error_code.h. But we need to provide some mechanism to support our internal mail_error::error_code.

dpotepalov replied:
>  when we use standard error_code we loose an error message form PostgreSQL

why? can we not communicate that message through a custom error_category?

thed636 replied:
> why? can we not communicate that message through a custom error_category?

Nope - error_category can not provide an ability for context depended information such as error messages from database which contains information about syntax error in query's text for example. So you can say "oh, you have a syntax error" with standard error_code, but you can not say "oh, you have a syntax error at column 45". "at column 45" is a context depended information, and there is no rooms for it in std or boos error_code. What's why we are using custom error_code in the meta access.

elsid commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+}
+
+
+struct dsn; // or just an alias to std::string?
```
Alias is enough.

elsid commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+};
+
+// Returns ConnectionProvider for specified dsn
+auto make_connection_provider(io_service&, dsn); 
```
Not only dsn, but config.

thed636 replied:
I'm totally agree with you.

elsid commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+	}
+}
+
+enum class parameters_transfer_mode {binary, text};
```
It is both the parameters and the result, may be just ```protocol_type```?

thed636 replied:
Yes, good idea!

thed636 commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+
+// Return ConnectionProvider for specified connection pool
+auto make_connection_provider(io_service& ios, connection_pool& pool) {
+	return [](auto&& handler) {
```
```return [&ios](auto&& handler) {``` of course.

thed636 commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+
+	template <typename ConnectionProvider, typename CompletionToken>
+	static auto execute(ConnectionProvider provider, query q, CompletionToken token);
+
```
I missed a ```prepare(...)``` function here.

elsid commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+* Should we allow to copy connection itself, or the object must be move-only?
+*/
+struct connection {
+	io_service& ios;
```
No tabs in this project, please.

thed636 replied:
I'll fix it. Didn't cate about tabs for the prototype. :)

elsid commented on ideas/single_connection_usage_with_timeouts.cpp:
```diff
@@ -2,12 +2,13 @@ boost::asio::io_service io_service;
     timeouts.request = 200ms;
-    apq::text_protocol::request(connection, cursor, timeouts, "SELECT pg_is_in_recovery()", yield);
-    std::cout << cursor.next(yield).at(0) << '\n';
+    if (apq::text_protocol::request(connection, apq::make_query("SELECT pg_is_in_recovery()", timeouts), row, yield)) {
```
Why timeouts are the part of query?

elsid replied:
What is the result of request, why we should use it as condition? Besides, there is always one row result in this example.

elsid commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+/**
+* Represents raw data row in case of request.
+*/
+class data_row;
```
Why not just ```row```?

thed636 replied:
I can't remember the actual reasons why not - let it be just ```row``` :)

elsid committed:
Use fetch in examples

elsid committed:
Example to show meaning of the fetch result

elsid committed:
Place cursor after query, and timeouts after query before cursor

elsid committed:
Add example with transaction_connection usage

thed636 commented on ideas/single_connection_usage_with_timeouts.cpp:
```diff
@@ -0,0 +1,15 @@
+    connection_timeouts timeouts;
+    timeouts.connect = 100ms;
+    timeouts.request = 200ms;
+    apq::text_protocol::request(connection, apq::make_query("SELECT pg_is_in_recovery()", timeouts), cursor, yield);
```
About timeouts. In this case we allow to specify custom timeouts for the query. It is useful feature and IMO it is better to provide it rather than force a user to make his (or hers) own solution.

elsid committed:
Bind timeouts to query

thed636 commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+
+
+template <parameters_transfer_mode prm>
+struct basic_protocol {
```
It looks like we do not need text protocol in case of binary protocol properly implemented. So we can simplify this part.

dpotepalov replied:
um... I tend to disagree. Text protocol is vastly simpler and can be always used as fallback in case something goes wrong with the binary (old server, new server, bug in some rarely used type etc). libpq is flexible enough to allow to choose representation for each query parameter and for the result also. Shouldn't we provide at least as much flexibility?

thed636 replied:
The text protocol and binary protocol are differing only in a way how to represent parameters. In case of binary protocol, the parameters represented as is in a binary way and marked with OIDs. In text protocol, you need to serialize all values as strings manually (including escaping which is not trivial in some cases and so on) point their types in a query mandatory and at the end pass them as a binary representation of strings. Do we need such a protocol support?

thed636 commented on ideas/apq.h:
```diff
@@ -0,0 +1,285 @@
+BOOST_FUSION_DEFINE_STRUCT((), (my_query_params), 
+	(std::string, id)
+	(int, sid)
+);
```
We need some type definition mechanism to allow map C++ types to database types through the OIDs and type names. So we need some traits system with easy macro for custom types definition.  We can predefine all the traits for geninue types like int, string and so on. E.g. for composite types we need something like BOOST_FUSION_DEFINE_STRUCT but with type traits definition inside, let it be something like this:

    APQ_DEFINE_COMPOSITE(<namespace>, <schema>, <type-name>, <composite-fields>);
    APQ_DEFINE_CUSTOM_TYPE(<namespace>, <schema>, <type-name>, <type-definition>);
    APQ_DEFINE_ENUM(<namespace>, <schema>, <type-name>, <enum-items>);

Another problem is UTF-8 string representation. Do we need to provide some custom type of validated UTF-8 or just use std::string?

Commit:
Add new ideas for query interface and type traits

It seems like a query may not be certain type rather that
concept with requirements and function customization point

thed636 commented on ideas/apq.h:
```diff
@@ -0,0 +1,362 @@
+* Oid acquisition for types must looks like (simplified):
+* auto oid = type_oid(v, [oid_map](const auto& v) {
+*         return oid_map.at(type_name(v));
+*     });
```
This case can be built with a compile-time check. It requires preparing something like boost::fusion::map with a key by type. So values of such "map" can be filled during connecting to database, or on error with invalidated oid.

elsid committed:
Rename query -> make_query

elsid committed:
Add examples with query builder and query conf

elsid committed:
Read query conf file

elsid committed:
Fix typo

elsid committed:
Add arguments type to query type
